client2/thin.go

```diff
@@ -0,0 +1,308 @@
+// along with a SURB so that you can later receive the reply along with the SURBID you choose.
+// This method of sending messages should be considered to be asynchronous because it does NOT actually wait until
+// the client daemon sends the message. Nor does it wait for a reply. The only blocking aspect to it's behavior is
+// merely blocking until the client daemon receives our request to send a message.
+func (t *ThinClient) SendMessage(surbID *[sConstants.SURBIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) error {
+	if surbID == nil {
+		return errors.New("surbID cannot be nil")
+	}
+
+	req := new(Request)
+	req.WithSURB = true
+	req.SURBID = surbID
+	req.IsSendOp = true
+	req.Payload = payload
+	req.DestinationIdHash = destNode
+	req.RecipientQueueID = destQueue
+	req.IsSendOp = true
+
+	blob, err := cbor.Marshal(req)
+	if err != nil {
+		return err
+	}
+	count, _, err := t.unixConn.WriteMsgUnix(blob, nil, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+	if count != len(blob) {
+		return fmt.Errorf("SendMessage error: wrote %d instead of %d bytes", count, len(blob))
+	}
+
+	return nil
+}
+
+func (t *ThinClient) readNextMessage() (*Response, error) {
+	buff := make([]byte, 65536)
+	msgLen, _, _, _, err := t.unixConn.ReadMsgUnix(buff, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	response := Response{}
+	err = cbor.Unmarshal(buff[:msgLen], &response)
+	if err != nil {
+		return nil, err
+	}
+	return &response, nil
+}
+
+// ResponseChan returns the channel that receives message responses
+// from the client daemon, of type ThinResponse.
+func (t *ThinClient) ResponseChan() chan ThinResponse {
+	return t.receivedCh
+}
+
+// ReceiveMessage blocks until a message is received.
+// Use ResponseChan instead if you want an async way to receive messages.
+func (t *ThinClient) ReceiveMessage() (*[sConstants.SURBIDLength]byte, []byte) {
+	resp := <-t.receivedCh
+	return resp.SURBID, resp.Payload
+}
+
+func (t *ThinClient) ARQSend(id *[MessageIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) error {
+	req := new(Request)
+	req.ID = id
+	req.WithSURB = true
+	req.IsARQSendOp = true
+	req.Payload = payload
+	req.DestinationIdHash = destNode
+	req.RecipientQueueID = destQueue
+
+	blob, err := cbor.Marshal(req)
+	if err != nil {
+		return err
+	}
+	count, _, err := t.unixConn.WriteMsgUnix(blob, nil, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+	if count != len(blob) {
+		return fmt.Errorf("SendMessage error: wrote %d instead of %d bytes", count, len(blob))
+	}
+
+	return nil
+}
+
+// ARQReceiveMessage blocks until a message is received.
+// Use ResponseChan instead if you want an async way to receive messages.
+func (t *ThinClient) ARQReceiveMessage() (*[MessageIDLength]byte, []byte) {
+	resp := <-t.receivedCh
+	return resp.ID, resp.Payload
+}
+
+func (t *ThinClient) BlockingSendReliableMessage(id *[MessageIDLength]byte, payload []byte, destNode *[32]byte, destQueue []byte) (reply []byte, err error) {
```

Comment:
this `id` thing could also use a longer name, so as not to confuse the requestID with the other IDs.
ie instead of `id` and `id2` I'd like to see `request_id` and `response_id`

Reply:
same id